icu4j/main/classes/core/src/com/ibm/icu/impl/number/DecimalQuantity_AbstractBCD.java

```diff
@@ -1033,12 +1018,19 @@ public String toPlainString() {
         if (precision == 0 || getMagnitude() < 0) {
             sb.append('0');
         }
-        for (int m = getUpperDisplayMagnitude(); m >= getLowerDisplayMagnitude(); m--) {
-            sb.append((char) ('0' + getDigit(m)));
-            if (m == 0)
-                sb.append('.');
+        if (exponent != 0) {
+            DecimalQuantity_AbstractBCD copy = (DecimalQuantity_AbstractBCD) this.createCopy();
```

Comment:
The reason I resorted to using a copy was because the precision seems to change between compact notation and scientific notation.  For the number 123456789, the precision in compact notation is 3, but in scientific notation is 7.  And that affects the bounds of the index iteration for getting digits (between `getUpperDisplayMagnitude()` and `getLowerDisplayMagnitude()`)

If I change the code to the following, all of the tests in `DecimalQuantityTest` appear to pass.  But it seems messy to me, and I can't tell if it's wrong (according to the semantic of some method) and/or unreasonable.

```
        if (exponent > 0) {
            int magnitudeStart = getUpperDisplayMagnitude();
            int magnitudeEnd = Math.min(getLowerDisplayMagnitude() - exponent - scale, magnitudeStart - precision + 1);
            for (int m = magnitudeStart; m >= magnitudeEnd; m--) {
                sb.append((char) ('0' + getDigit(m)));
                if (m == -exponent)
                    sb.append('.');
            }
            return sb.toString();
        } else if (exponent < 0) {
            int magnitudeStart = getUpperDisplayMagnitude() - exponent - scale;
            int magnitudeEnd = Math.min(getLowerDisplayMagnitude(), magnitudeStart - precision + 1);
            for (int m = magnitudeStart; m >= magnitudeEnd; m--) {
                sb.append((char) ('0' + getDigit(m)));
                if (m == -exponent)
                    sb.append('.');
            }
            return sb.toString();
        } else {
```

Reply:
You can try implementing this method without using `get*DisplayMagnitude`, like `toScientificString()`.  I think the way you found to fix `toScientificString()` is really nice.

Reply:
Done.

Reply:
I re-implemented `toPlainString()` and pushed it to your branch.  18 lines of code and no object copies.